src/model_theory/rings/language.lean

```diff
@@ -0,0 +1,106 @@
+/-
+Copyright (c) 2022 Joseph Hua. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Hua
+-/
+import model_theory.terms_and_formulas
+
+/-!
+# The language of rings
+
+## Main Definitions
+* A `first_order.language.ring.L` defines the language of rings,
+  which consists of 0,1,-,+,*.
+-/
+universes u v u'
+
+namespace first_order
+namespace language
+namespace ring
+
+/-- The constant symbols in ring.L -/
+inductive consts : Type*
+| zero : consts
+| one : consts
+
+/-- The unary function symbols in ring.L-/
+inductive unaries : Type*
+| neg : unaries
+
+/-- The binary function symbols in ring.L-/
+inductive binaries : Type*
+| add : binaries
+| mul : binaries
+
+/-- All function symbols in ring.langauge-/
+def functions : ℕ → Type*
+| 0 := consts
+| 1 := unaries
+| 2 := binaries
+| (n + 3) := pempty
+
+instance : inhabited consts := ⟨ consts.zero ⟩
+instance : inhabited unaries := ⟨ unaries.neg ⟩
+instance : inhabited binaries := ⟨ binaries.add ⟩
+instance : inhabited (functions 0) := ⟨ consts.zero ⟩
+
+/-- The language of rings -/
+def L : language.{u v} :=
+{ functions := functions,
+  relations := λ n, pempty }
+
+variable {α : Type u'}
+
+@[simp] instance : has_zero (L.term α) := ⟨ @func L _ 0 consts.zero fin_zero_elim ⟩
+
+@[simp] instance : has_one (L.term α) := ⟨ @func L _ 0 consts.one fin_zero_elim ⟩
+
+@[simp] instance : has_neg (L.term α) := ⟨ λ x, @func L _ 1 unaries.neg (λ _, x) ⟩
+
+@[simp] instance : has_add (L.term α) :=
+⟨ λ x y, @func L _ 2 binaries.add (λ i, match i with | ⟨0, _⟩ := x | ⟨n+1, _⟩ := y end) ⟩
+
+@[simp] instance : has_mul (L.term α) :=
+⟨ λ x y, @func L _ 2 binaries.mul (λ i, match i with | ⟨0, _⟩ := x | ⟨n+1, _⟩ := y end) ⟩
+
+instance : has_pow (L.term α) ℕ := ⟨ λ t n, npow_rec n t ⟩
+
+@[simp] lemma pow_zero (t : L.term α) : t ^ 0 = 1 := rfl
+@[simp] lemma pow_succ {n} (t : L.term α) : t ^ (n + 1) = t * t ^ n := rfl
+
+lemma fin_zero_elim_uniq (f : fin 0 → α) : f = fin_zero_elim := subsingleton.elim _ _
+
+lemma fin_one_eta (f : fin 1 → α) : (λ (_ : fin 1), f 0) = f :=
+funext (λ x, by {simp [unique.eq_default x] })
+
+lemma not_succ_succ_lt_two {n : ℕ} (h : n.succ.succ < 2) : false :=
+by {apply nat.not_lt_zero n, repeat {rw nat.succ_lt_succ_iff at h}, exact h }
+
+/-- Part of the definition of ring_term_rec -/
+@[simp] def term_rec_functions {C : L.term α → Sort*} (h0 : C 0)
```

Comment:
I should read more of your off-mathlib project first, but I'm not sure yet if this recursor saves effort rather than defining your functions later with the equation compiler directly. If you use the equation compiler directly, none of this casting will be required.

Would it perhaps be a good compromise to define this only for when `C` is constant? Then the casting will still not be required.

Reply:
The way I see it, the really important thing is to define a map to `mv_polynomial \a`, and then to prove (presumably using the `induction` tactic) that in any reasonable structure, `t.realize x` is given by evaluating said polynomial.

Reply:
> I should read more of your off-mathlib project first, but I'm not sure yet if this recursor saves effort rather than defining your functions later with the equation compiler directly. If you use the equation compiler directly, none of this casting will be required.
> 
> Would it perhaps be a good compromise to define this only for when `C` is constant? Then the casting will still not be required.

What do you mean when `C` is constant?

Reply:
I've removed this part for now, and I can introduce it in a later PR if need be.